### PR TITLE
Reimplemented sys.babelfish_concat_wrapper in C to improve the performance

### PR DESCRIPTION
**Pull request by ParikshitSarode**
### Description -  Reimplemented sys.babelfish_concat_wrapper in C to improve the performance

1. The '+' operator it took more time to complete the query for concatenation operation.
2. This was due to the function babelfish_concat_wrapper was implemented in SQL.
3. In this change, the babelfish_concat_wrapper function is implemented in C.
4. This improves the '+' operator query execution time by 55%

Signed-off-by: Parikshit Sarode [pasarode@amazon.com](mailto:pasarode@amazon.com)

### Issues Resolved
Task: BABEL-4299

### Check List
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 1]
```diff
@@ -181,13 +182,54 @@ const char *bbf_servicename = "MSSQLSERVER";
...
+		if(PG_ARGISNULL(0) || PG_ARGISNULL(1))
+  		{
+  		  	PG_RETURN_NULL(); // If any is NULL, return NULL
+   		}
+	}
+	else
+	{
+		if (PG_ARGISNULL(0) && PG_ARGISNULL(1))
+    	{
+     	   PG_RETURN_NULL(); // If both are NULL, return NULL
```
Comment (forestkeeper) [inline contrib/babelfishpg_tsql/runtime/functions.c]: Would you mind adjust the interval format, thanks

**Review by forestkeeper — COMMENTED** [review]
Overall looks good, please adjust the tab for the if conditions

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 2]
```diff
@@ -181,13 +182,56 @@ const char *bbf_servicename = "MSSQLSERVER";
...
+	arg1 = PG_GETARG_TEXT_PP(0);
+	arg2 = PG_GETARG_TEXT_PP(1);
+	arg1_size = VARSIZE_ANY_EXHDR(arg1);
+	arg2_size = VARSIZE_ANY_EXHDR(arg2);
+	new_text_size = arg1_size + arg2_size + VARHDRSZ;
+	new_text = (text *) palloc(new_text_size);
+
+	SET_VARSIZE(new_text, new_text_size);
+	memcpy(VARDATA(new_text), VARDATA_ANY(arg1), arg1_size);
+	memcpy(VARDATA(new_text) + arg1_size, VARDATA_ANY(arg2), arg2_size);
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/runtime/functions.c]: seems like this implementation is inspired by `text_catenate` but lets put assert here to check arg1_size and arg2_size > 0? what would happen if empty string is passed? and is there existing tests cases to validate this change?

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 3]
```diff
@@ -181,13 +182,64 @@ const char *bbf_servicename = "MSSQLSERVER";
...
+		else if (first_param)
+		{
+			PG_RETURN_TEXT_P(PG_GETARG_TEXT_PP(1)); // If only the first string is NULL, return the second string
+		}
+	}
+	arg1 = PG_GETARG_TEXT_PP(0);
+	arg2 = PG_GETARG_TEXT_PP(1);
+	arg1_size = VARSIZE_ANY_EXHDR(arg1);
+	arg2_size = VARSIZE_ANY_EXHDR(arg2);
+	new_text_size = arg1_size + arg2_size + VARHDRSZ;
```
Comment (forestkeeper) [inline contrib/babelfishpg_tsql/runtime/functions.c:224]: Does the new size has a maximum limit ?

Comment (ParikshitSarode) [reply contrib/babelfishpg_tsql/runtime/functions.c:224]: updated

**Review by Deepesh125 — APPROVED** [review]

**Pull request merged**